File: backend/tests/run_tests.py
#!/usr/bin/env python3
"""
Comprehensive test runner for Hubstaff Clone API
This script runs all tests and provides detailed reporting
"""
import asyncio
import subprocess
import sys
import os
from pathlib import Path
import time
import httpx
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Color codes for terminal output
class Colors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

def print_colored(message, color=Colors.ENDC):
    """Print colored message"""
    print(f"{color}{message}{Colors.ENDC}")

def print_header(message):
    """Print header message"""
    print_colored(f"\n{'='*60}", Colors.HEADER)
    print_colored(f"{message}", Colors.HEADER)
    print_colored(f"{'='*60}", Colors.HEADER)

def print_success(message):
    """Print success message"""
    print_colored(f"✅ {message}", Colors.OKGREEN)

def print_error(message):
    """Print error message"""
    print_colored(f"❌ {message}", Colors.FAIL)

def print_warning(message):
    """Print warning message"""
    print_colored(f"⚠️  {message}", Colors.WARNING)

def print_info(message):
    """Print info message"""
    print_colored(f"ℹ️  {message}", Colors.OKBLUE)

async def check_server_health():
    """Check if the server is running and healthy"""
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            response = await client.get("http://localhost:8001/health")
            if response.status_code == 200:
                print_success("Server is running and healthy")
                return True
            else:
                print_error(f"Server health check failed: {response.status_code}")
                return False
    except Exception as e:
        print_error(f"Cannot connect to server: {e}")
        print_info("Make sure the server is running: uvicorn server:app --host 0.0.0.0 --port 8001")
        return False

def run_pytest_tests():
    """Run pytest tests with detailed output"""
    print_header("Running Backend API Tests")
    
    # Change to backend directory
    backend_dir = Path(__file__).parent.parent
    os.chdir(backend_dir)
    
    # Run pytest with verbose output
    cmd = [
        sys.executable, "-m", "pytest",
        "tests/",
        "-v",
        "--tb=short",
        "--color=yes",
        "--durations=10",
        "-x"  # Stop on first failure
    ]
    
    print_info(f"Running command: {' '.join(cmd)}")
    
    try:
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        # Print stdout
        if result.stdout:
            print(result.stdout)
        
        # Print stderr
        if result.stderr:
            print_colored("STDERR:", Colors.WARNING)
            print(result.stderr)
        
        if result.returncode == 0:
            print_success("All tests passed!")
            return True
        else:
            print_error(f"Tests failed with return code: {result.returncode}")
            return False
            
    except Exception as e:
        print_error(f"Failed to run pytest: {e}")
        return False

def run_manual_endpoint_tests():
    """Run manual endpoint tests using curl"""
    print_header("Running Manual Endpoint Tests")
    
    # Test basic endpoints without authentication
    basic_endpoints = [
        ("GET", "http://localhost:8001/health", "Health Check"),
        ("GET", "http://localhost:8001/docs", "API Documentation"),
    ]
    
    for method, url, description in basic_endpoints:
        try:
            cmd = ["curl", "-s", "-w", "%{http_code}", "-o", "/dev/null", url]
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=10)
            
            status_code = result.stdout.strip()
            if status_code == "200":
                print_success(f"{description}: {status_code}")
            else:
                print_warning(f"{description}: {status_code}")
                
        except subprocess.TimeoutExpired:
            print_error(f"{description}: Timeout")
        except Exception as e:
            print_error(f"{description}: {e}")

def check_test_requirements():
    """Check if all test requirements are installed"""
    print_header("Checking Test Requirements")
    
    required_packages = [
        "pytest",
        "httpx",
        "websockets",
        "pytest-asyncio"
    ]
    
    missing_packages = []
    
    for package in required_packages:
        try:
            __import__(package.replace("-", "_"))
            print_success(f"{package} is installed")
        except ImportError:
            print_error(f"{package} is missing")
            missing_packages.append(package)
    
    if missing_packages:
        print_warning("Install missing packages with:")
        print_info(f"pip install {' '.join(missing_packages)}")
        return False
    
    return True

def generate_test_report():
    """Generate a test report"""
    print_header("Test Report")
    
    report_data = {
        "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
        "server_status": "Unknown",
        "test_results": "Unknown"
    }
    
    # Save report to file
    report_file = Path("test_report.txt")
    with open(report_file, "w") as f:
        f.write(f"Hubstaff Clone API Test Report\n")
        f.write(f"Generated: {report_data['timestamp']}\n")
        f.write(f"{'='*50}\n\n")
        f.write(f"Server Status: {report_data['server_status']}\n")
        f.write(f"Test Results: {report_data['test_results']}\n")
    
    print_info(f"Test report saved to: {report_file}")

async def main():
    """Main test runner function"""
    print_header("Hubstaff Clone API Test Suite")
    
    # Check requirements
    if not check_test_requirements():
        print_error("Missing test requirements. Please install them first.")
        return 1
    
    # Check server health
    server_healthy = await check_server_health()
    if not server_healthy:
        print_error("Server is not healthy. Cannot run API tests.")
        return 1
    
    # Run manual tests
    run_manual_endpoint_tests()
    
    # Run pytest tests
    tests_passed = run_pytest_tests()
    
    # Generate report
    generate_test_report()
    
    if tests_passed:
        print_success("All tests completed successfully!")
        return 0
    else:
        print_error("Some tests failed. Please check the output above.")
        return 1

if __name__ == "__main__":
    try:
        exit_code = asyncio.run(main())
        sys.exit(exit_code)
    except KeyboardInterrupt:
        print_warning("\nTests interrupted by user")
        sys.exit(1)
    except Exception as e:
        print_error(f"Test runner failed: {e}")
        sys.exit(1)

File: frontend/tests/runTests.js
#!/usr/bin/env node

/**
 * Frontend test runner with detailed reporting
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

// Colors for console output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
};

function colorize(text, color) {
  return `${colors[color]}${text}${colors.reset}`;
}

function printHeader(message) {
  console.log('\n' + '='.repeat(60));
  console.log(colorize(message, 'cyan'));
  console.log('='.repeat(60));
}

function printSuccess(message) {
  console.log(colorize(`✅ ${message}`, 'green'));
}

function printError(message) {
  console.log(colorize(`❌ ${message}`, 'red'));
}

function printWarning(message) {
  console.log(colorize(`⚠️  ${message}`, 'yellow'));
}

function printInfo(message) {
  console.log(colorize(`ℹ️  ${message}`, 'blue'));
}

async function checkDependencies() {
  printHeader('Checking Dependencies');
  
  const requiredDeps = [
    '@testing-library/react',
    '@testing-library/jest-dom',
    '@testing-library/user-event'
  ];
  
  const packageJsonPath = path.join(__dirname, '..', 'package.json');
  
  if (!fs.existsSync(packageJsonPath)) {
    printError('package.json not found');
    return false;
  }
  
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
  const allDeps = { ...packageJson.dependencies, ...packageJson.devDependencies };
  
  let allInstalled = true;
  
  requiredDeps.forEach(dep => {
    if (allDeps[dep]) {
      printSuccess(`${dep} is installed`);
    } else {
      printError(`${dep} is missing`);
      allInstalled = false;
    }
  });
  
  if (!allInstalled) {
    printWarning('Install missing dependencies with:');
    printInfo('npm install --save-dev @testing-library/react @testing-library/jest-dom @testing-library/user-event');
  }
  
  return allInstalled;
}

function runTests() {
  return new Promise((resolve) => {
    printHeader('Running Frontend Tests');
    
    const testProcess = spawn('npm', ['test', '--', '--watchAll=false', '--verbose'], {
      stdio: 'inherit',
      cwd: path.join(__dirname, '..'),
      shell: true
    });
    
    testProcess.on('close', (code) => {
      if (code === 0) {
        printSuccess('All tests passed!');
      } else {
        printError(`Tests failed with exit code: ${code}`);
      }
      resolve(code === 0);
    });
    
    testProcess.on('error', (error) => {
      printError(`Failed to run tests: ${error.message}`);
      resolve(false);
    });
  });
}

function runComponentTests() {
  return new Promise((resolve) => {
    printHeader('Running Component Tests');
    
    const testProcess = spawn('npm', ['test', '--', 'tests/components', '--watchAll=false', '--verbose'], {
      stdio: 'inherit',
      cwd: path.join(__dirname, '..'),
      shell: true
    });
    
    testProcess.on('close', (code) => {
      resolve(code === 0);
    });
    
    testProcess.on('error', (error) => {
      printError(`Failed to run component tests: ${error.message}`);
      resolve(false);
    });
  });
}

function runIntegrationTests() {
  return new Promise((resolve) => {
    printHeader('Running Integration Tests');
    
    const testProcess = spawn('npm', ['test', '--', 'tests/integration', '--watchAll=false', '--verbose'], {
      stdio: 'inherit',
      cwd: path.join(__dirname, '..'),
      shell: true
    });
    
    testProcess.on('close', (code) => {
      resolve(code === 0);
    });
    
    testProcess.on('error', (error) => {
      printError(`Failed to run integration tests: ${error.message}`);
      resolve(false);
    });
  });
}

async function runCoverage() {
  return new Promise((resolve) => {
    printHeader('Running Coverage Report');
    
    const coverageProcess = spawn('npm', ['test', '--', '--coverage', '--watchAll=false'], {
      stdio: 'inherit',
      cwd: path.join(__dirname, '..'),
      shell: true
    });
    
    coverageProcess.on('close', (code) => {
      if (code === 0) {
        printSuccess('Coverage report generated');
        printInfo('Coverage report available in coverage/ directory');
      } else {
        printWarning('Coverage report generation failed');
      }
      resolve(code === 0);
    });
    
    coverageProcess.on('error', (error) => {
      printError(`Failed to run coverage: ${error.message}`);
      resolve(false);
    });
  });
}

function generateTestReport() {
  printHeader('Test Report');
  
  const report = {
    timestamp: new Date().toISOString(),
    testResults: 'See above output for detailed results'
  };
  
  const reportPath = path.join(__dirname, 'test-report.json');
  fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));
  
  printInfo(`Test report saved to: ${reportPath}`);
}

async function main() {
  try {
    printHeader('Frontend Test Suite');
    
    // Check dependencies
    const depsOk = await checkDependencies();
    if (!depsOk) {
      printError('Missing dependencies. Please install them first.');
      process.exit(1);
    }
    
    // Parse command line arguments
    const args = process.argv.slice(2);
    
    if (args.includes('--help') || args.includes('-h')) {
      console.log(`
Usage: node runTests.js [options]

Options:
  --components     Run only component tests
  --integration    Run only integration tests
  --coverage       Run tests with coverage report
  --help, -h       Show this help message

Examples:
  node runTests.js                    # Run all tests
  node runTests.js --components       # Run component tests only
  node runTests.js --coverage         # Run all tests with coverage
      `);
      return;
    }
    
    let success = true;
    
    if (args.includes('--components')) {
      success = await runComponentTests();
    } else if (args.includes('--integration')) {
      success = await runIntegrationTests();
    } else if (args.includes('--coverage')) {
      success = await runCoverage();
    } else {
      // Run all tests
      const componentSuccess = await runComponentTests();
      const integrationSuccess = await runIntegrationTests();
      success = componentSuccess && integrationSuccess;
      
      if (success) {
        // Run coverage if all tests pass
        await runCoverage();
      }
    }
    
    // Generate report
    generateTestReport();
    
    if (success) {
      printSuccess('All tests completed successfully!');
      process.exit(0);
    } else {
      printError('Some tests failed. Please check the output above.');
      process.exit(1);
    }
    
  } catch (error) {
    printError(`Test runner failed: ${error.message}`);
    process.exit(1);
  }
}

// Handle process interruption
process.on('SIGINT', () => {
  printWarning('\nTests interrupted by user');
  process.exit(1);
});

if (require.main === module) {
  main();
}